backend/commons/src/main/java/org/sonarsource/sonarlint/core/commons/util/GitUtils.java

```diff
@@ -0,0 +1,89 @@
+/*
+ * SonarLint Core - Commons
+ * Copyright (C) 2016-2024 SonarSource SA
+ * mailto:info AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonarsource.sonarlint.core.commons.util;
+
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.IOException;
+import java.net.URI;
+import java.nio.file.Path;
+import java.util.List;
+import java.util.Optional;
+import java.util.stream.Collectors;
+import org.eclipse.jgit.ignore.FastIgnoreRule;
+import org.eclipse.jgit.ignore.IgnoreNode;
+import org.eclipse.jgit.lib.Repository;
+import org.sonarsource.sonarlint.core.commons.log.SonarLintLogger;
+
+import static org.eclipse.jgit.lib.Constants.GITIGNORE_FILENAME;
+import static org.sonarsource.sonarlint.core.commons.util.gitblame.GitBlameUtils.buildGitRepository;
+
+public class GitUtils {
+
+  private static final SonarLintLogger LOG = SonarLintLogger.get();
+
+  private GitUtils() {
+    // Utility class
+  }
+
+  /**
+   * Assumes the supplied {@param baseDir} or some of its parents is a git repository.
+   * The method checks {@param filePathsToAnalyze} against patterns defined in the .gitignore file and returns
+   * those that are not ignored.
+   * If error occurs during execution then all {@param filePathsToAnalyze} are considered as not ignored.
+   */
+  public static List<URI> filterOutIgnoredFiles(Path baseDir, List<URI> filePathsToAnalyze) {
```

Comment:
I think `filterOutIgnoredFiles` is a bit too specific as a utility method. What about introducing a new class like `SonarLintGitIgnore` similar to `SonarLintBlameResult`?

```java
public class SonarLintGitIgnore {
  private final IgnoreNode ignoreNode;
  private final Path gitRepoRelativeProjectBaseDir;

  public SonarLintGitIgnore(IgnoreNode ignoreNode, Path gitRepoRelativeProjectBaseDir) {
    this.ignoreNode = ignoreNode;
    this.gitRepoRelativeProjectBaseDir = gitRepoRelativeProjectBaseDir;
  }

  public boolean isIgnored(String entryPath, boolean isDirectory) {
    List<FastIgnoreRule> rules = ignoreNode.getRules();
    // Parse rules in the reverse order that they were read because later rules have higher priority
    for (var i = rules.size() - 1; i > -1; i--) {
      var rule = rules.get(i);
      if (rule.isMatch(entryPath, isDirectory)) {
        return rule.getResult();
      }
    }
    return false;
  }
}
 ```

In this way:
- we could provide a more general git ignore utility, including the directory check
- encapsulate relative path logic
- chance to check one file
- keep the filtering business logic in the analysis service

Wdyt?

Reply:
Fixed in the last commit (34ae30b02d40dedea6c77724771076313aa9ef49)